File: modules/agent/src/controllers/conversation_controller.py
"""Controller layer for conversation routing to services."""

from typing import List
from uuid import UUID

from fastapi import Request

from lib.decorators import handle_http_exceptions
from schemas.conversation import ConversationUpdate
from services import conversation_service as service


@handle_http_exceptions
async def list_conversations(
    request: Request,
    limit: int = 50,
    include_archived: bool = False,
) -> List[dict]:
    """List conversations for the current user."""
    user_id = request.state.user_id
    tenant_id = request.state.tenant_id
    conversations = await service.list_conversations(
        user_id=user_id,
        tenant_id=tenant_id,
        limit=limit,
        include_archived=include_archived,
    )
    return [c.model_dump() for c in conversations]


@handle_http_exceptions
async def list_tenant_conversations(
    request: Request,
    search: str = None,
    limit: int = 100,
    offset: int = 0,
    include_archived: bool = False,
) -> dict:
    """List all conversations for the tenant (with search and pagination)."""
    tenant_id = request.state.tenant_id
    conversations, total = await service.list_tenant_conversations(
        tenant_id=tenant_id,
        search=search,
        limit=limit,
        offset=offset,
        include_archived=include_archived,
    )
    return {
        "data": [c.model_dump() for c in conversations],
        "total": total,
        "limit": limit,
        "offset": offset,
    }


@handle_http_exceptions
async def get_conversation(request: Request, thread_id: UUID) -> dict:
    """Get a conversation by thread_id."""
    result = await service.get_conversation_with_messages(thread_id)
    return result.model_dump()


@handle_http_exceptions
async def update_title(request: Request, thread_id: UUID, data: ConversationUpdate) -> dict:
    """Update conversation title."""
    conversation = await service.update_title(thread_id, data.title)
    return conversation.model_dump()


@handle_http_exceptions
async def archive_conversation(request: Request, thread_id: UUID) -> dict:
    """Archive a conversation."""
    await service.archive_conversation(thread_id)
    return {"status": "archived"}


@handle_http_exceptions
async def unarchive_conversation(request: Request, thread_id: UUID) -> dict:
    """Unarchive a conversation."""
    await service.unarchive_conversation(thread_id)
    return {"status": "unarchived"}


@handle_http_exceptions
async def delete_permanent(request: Request, thread_id: UUID) -> dict:
    """Permanently delete an archived conversation."""
    await service.delete_permanent(thread_id)
    return {"status": "deleted"}

File: modules/agent/internal/controllers/conversation_controller.go
package controllers

import (
	"encoding/json"
	"net/http"
	"strconv"

	"github.com/go-chi/chi/v5"
	"agent/internal/middleware"
	"agent/internal/services"
)

// ConversationController handles conversation HTTP requests
type ConversationController struct {
	convService *services.ConversationService
}

// NewConversationController creates a new conversation controller
func NewConversationController(convService *services.ConversationService) *ConversationController {
	return &ConversationController{convService: convService}
}

// GetConversations handles GET /conversations
func (c *ConversationController) GetConversations(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "unauthorized")
		return
	}

	var tenantID *int64
	if tid, ok := middleware.GetTenantID(r.Context()); ok {
		tenantID = &tid
	}

	includeArchived := r.URL.Query().Get("include_archived") == "true"
	limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))
	if limit <= 0 {
		limit = 50
	}

	results, err := c.convService.GetConversations(userID, tenantID, includeArchived, limit)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, results)
}

// GetTenantConversations handles GET /conversations/all
func (c *ConversationController) GetTenantConversations(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant required")
		return
	}

	search := r.URL.Query().Get("search")
	includeArchived := r.URL.Query().Get("include_archived") == "true"

	limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))
	if limit <= 0 {
		limit = 100
	}

	offset, _ := strconv.Atoi(r.URL.Query().Get("offset"))

	results, err := c.convService.GetTenantConversations(tenantID, search, includeArchived, limit, offset)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, results)
}

// GetConversation handles GET /conversations/{thread_id}
func (c *ConversationController) GetConversation(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "unauthorized")
		return
	}

	threadID := chi.URLParam(r, "thread_id")
	if threadID == "" {
		writeError(w, http.StatusBadRequest, "thread_id required")
		return
	}

	result, err := c.convService.GetConversation(threadID, userID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	if result == nil {
		writeError(w, http.StatusNotFound, "conversation not found")
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// UpdateConversationTitle handles PATCH /conversations/{thread_id}
func (c *ConversationController) UpdateConversationTitle(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "unauthorized")
		return
	}

	threadID := chi.URLParam(r, "thread_id")
	if threadID == "" {
		writeError(w, http.StatusBadRequest, "thread_id required")
		return
	}

	var input struct {
		Title string `json:"title"`
	}
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	result, err := c.convService.UpdateTitle(threadID, userID, input.Title)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	if result == nil {
		writeError(w, http.StatusNotFound, "conversation not found")
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// ArchiveConversation handles DELETE /conversations/{thread_id}
func (c *ConversationController) ArchiveConversation(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "unauthorized")
		return
	}

	threadID := chi.URLParam(r, "thread_id")
	if threadID == "" {
		writeError(w, http.StatusBadRequest, "thread_id required")
		return
	}

	if err := c.convService.ArchiveConversation(threadID, userID); err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, map[string]bool{"success": true})
}

// UnarchiveConversation handles POST /conversations/{thread_id}/unarchive
func (c *ConversationController) UnarchiveConversation(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "unauthorized")
		return
	}

	threadID := chi.URLParam(r, "thread_id")
	if threadID == "" {
		writeError(w, http.StatusBadRequest, "thread_id required")
		return
	}

	if err := c.convService.UnarchiveConversation(threadID, userID); err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, map[string]bool{"success": true})
}

// DeleteConversationPermanent handles DELETE /conversations/{thread_id}/permanent
func (c *ConversationController) DeleteConversationPermanent(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "unauthorized")
		return
	}

	threadID := chi.URLParam(r, "thread_id")
	if threadID == "" {
		writeError(w, http.StatusBadRequest, "thread_id required")
		return
	}

	if err := c.convService.DeleteConversation(threadID, userID); err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, map[string]bool{"success": true})
}
